File: src/services/GrokService.js
import OpenAI from 'openai';
import dotenv from 'dotenv';
import configService from './ConfigService.js';

dotenv.config();

const DEMO_RESET_KEYWORDS = ['menu', 'demos', 'cambiar', 'ejemplos', 'volver', 'regresar', '0'];

class GrokService {
  constructor() {
    console.log('🔧 Inicializando GrokService...');
    console.log('📌 API Key presente:', !!process.env.XAI_API_KEY);

    this.client = new OpenAI({
      apiKey: process.env.XAI_API_KEY,
      baseURL: 'https://api.x.ai/v1'
    });

    this.systemPrompt = configService.getSystemPrompt();
    this.conversations = {};
    this.userMenuState = {}; // Track if user has seen menu
    this.userDemoState = {}; // Track per-user demo state { flowId, prompt }
    console.log('✅ GrokService inicializado correctamente\n');
  }

  updateSystemPrompt(newPrompt) {
    this.systemPrompt = newPrompt;
    console.log('✅ System prompt actualizado en GrokService');
  }

  shouldShowMenu(userId) {
    const currentFlow = configService.getCurrentFlow();
    const flowData = configService.getFlowData(currentFlow);
    
    // Show menu if: flow has menu, and user hasn't seen it yet
    if (flowData && flowData.hasMenu && flowData.menuConfig) {
      if (!this.userMenuState[userId]) {
        return true;
      }
    }
    return false;
  }

  buildMenuMessage(menuConfig) {
    let message = menuConfig.welcomeMessage + '\n\n';
    
    menuConfig.options.forEach((option, index) => {
      message += `${index + 1}. ${option.label}\n`;
    });
    
    message += '\n' + menuConfig.footerMessage;
    return message;
  }

  handleMenuSelection(userId, userMessage, menuConfig) {
    const selection = parseInt(userMessage.trim());

    if (isNaN(selection) || selection < 1 || selection > menuConfig.options.length) {
      return null;
    }

    const selectedOption = menuConfig.options[selection - 1];
    return {
      response: selectedOption.response,
      demoFlowId: selectedOption.demoFlowId
    };
  }

  async getResponse(userId, userMessage) {
    console.log('\n📨 Nueva solicitud a Grok:');
    console.log('  Usuario:', userId);
    console.log('  Mensaje:', userMessage);

    try {
      const currentFlow = configService.getCurrentFlow();
      const flowData = configService.getFlowData(currentFlow);
      const lowerMsg = userMessage.trim().toLowerCase();

      // Handle demo reset keywords — return to demo menu
      if (this.userDemoState[userId] && DEMO_RESET_KEYWORDS.includes(lowerMsg)) {
        delete this.userDemoState[userId];
        this.conversations[userId] = [];
        delete this.userMenuState[userId];
        if (flowData && flowData.hasMenu && flowData.menuConfig) {
          this.userMenuState[userId] = true;
          return this.buildMenuMessage(flowData.menuConfig);
        }
      }

      // Check if we should show menu
      if (this.shouldShowMenu(userId)) {
        const menuConfig = configService.getMenuForFlow(currentFlow);

        if (menuConfig) {
          this.userMenuState[userId] = true;
          return this.buildMenuMessage(menuConfig);
        }
      }

      // Check if this is a menu selection (with demo flow switching)
      if (flowData && flowData.hasMenu && flowData.menuConfig) {
        const result = this.handleMenuSelection(userId, userMessage, flowData.menuConfig);
        if (result) {
          if (result.demoFlowId) {
            const demoFlow = configService.getFlowData(result.demoFlowId);
            if (demoFlow) {
              this.userDemoState[userId] = { flowId: result.demoFlowId, prompt: demoFlow.prompt };
              this.conversations[userId] = [];
              console.log(`  🎭 Demo activada para ${userId}: ${result.demoFlowId}`);
            }
          }
          console.log('  📋 Respuesta de menú seleccionada');
          return result.response;
        }
      }

      // Normal AI response
      if (!this.conversations[userId]) {
        this.conversations[userId] = [];
        console.log('  📝 Nueva conversación creada');
      }

      this.conversations[userId].push({
        role: 'user',
        content: userMessage
      });

      if (this.conversations[userId].length > 20) {
        this.conversations[userId] = this.conversations[userId].slice(-20);
        console.log('  ♻️  Historial recortado a 20 mensajes');
      }

      // Use per-user demo prompt if in demo mode, otherwise global prompt
      const currentPrompt = this.userDemoState[userId]
        ? this.userDemoState[userId].prompt
        : (this.systemPrompt || configService.getSystemPrompt());

      const demoLabel = this.userDemoState[userId]
        ? ` [DEMO: ${this.userDemoState[userId].flowId}]`
        : '';
      console.log('  📤 Enviando a Grok API...');
      console.log(`  🤖 Modelo: grok-4-fast-reasoning${demoLabel}`);
      console.log('  💬 Mensajes en contexto:', this.conversations[userId].length);

      const completion = await this.client.chat.completions.create({
        model: 'grok-4-fast-reasoning',
        messages: [
          { role: 'system', content: currentPrompt },
          ...this.conversations[userId]
        ],
        temperature: 0.7,
        max_tokens: 1000
      });

      console.log('  ✅ Respuesta recibida de Grok');

      const assistantMessage = completion.choices[0].message.content;
      console.log('  📝 Respuesta:', assistantMessage.substring(0, 100) + '...');

      this.conversations[userId].push({
        role: 'assistant',
        content: assistantMessage
      });

      return assistantMessage;

    } catch (error) {
      console.error('\n❌ ERROR EN GROK API:');
      console.error('  Mensaje:', error.message);
      console.error('  Status:', error.status);

      return 'Disculpa, hubo un error técnico. ¿Puedes intentar de nuevo?';
    }
  }

  clearConversation(userId) {
    this.conversations[userId] = [];
    delete this.userMenuState[userId];
    delete this.userDemoState[userId];
    console.log('🔄 Conversación reiniciada para:', userId);
  }

  getConversationHistory(userId) {
    return this.conversations[userId] || [];
  }
}

export default new GrokService();

File: backend/app/services/grok_service.py
"""Grok AI service for generating responses."""
from typing import Dict, List, Optional
from openai import OpenAI
from ..config import get_settings
from .config_service import config_service


class GrokService:
    """Service for interacting with Grok AI API."""

    def __init__(self):
        settings = get_settings()
        print("Initializing GrokService...")
        print(f"API Key present: {bool(settings.xai_api_key)}")

        self.client = OpenAI(
            api_key=settings.xai_api_key,
            base_url="https://api.x.ai/v1"
        ) if settings.xai_api_key else None

        self.system_prompt = config_service.get_system_prompt()
        self.conversations: Dict[str, List[Dict[str, str]]] = {}
        self.user_menu_state: Dict[str, bool] = {}

        if self.client:
            print("GrokService initialized successfully")
        else:
            print("WARNING: GrokService initialized without API key")

    def update_system_prompt(self, new_prompt: str) -> None:
        """Update the system prompt."""
        self.system_prompt = new_prompt
        print("System prompt updated in GrokService")

    def should_show_menu(self, user_id: str) -> bool:
        """Check if menu should be shown to user."""
        current_flow = config_service.get_current_flow()
        flow_data = config_service.get_flow_data(current_flow)

        if flow_data and flow_data.get("has_menu") and flow_data.get("menu_config"):
            if user_id not in self.user_menu_state:
                return True
        return False

    def build_menu_message(self, menu_config: Dict) -> str:
        """Build menu message from config."""
        message = menu_config.get("welcome_message", "") + "\n\n"

        for i, option in enumerate(menu_config.get("options", []), 1):
            message += f"{i}. {option.get('label', '')}\n"

        message += "\n" + menu_config.get("footer_message", "")
        return message

    def handle_menu_selection(
        self,
        user_id: str,
        user_message: str,
        menu_config: Dict
    ) -> Optional[str]:
        """Handle menu selection from user."""
        try:
            selection = int(user_message.strip())
            options = menu_config.get("options", [])

            if selection < 1 or selection > len(options):
                return None

            selected_option = options[selection - 1]
            return selected_option.get("response")
        except ValueError:
            return None

    async def get_response(self, user_id: str, user_message: str) -> str:
        """Get AI response for user message."""
        print(f"\nNew request to Grok:")
        print(f"  User: {user_id}")
        print(f"  Message: {user_message}")

        if not self.client:
            return "Lo siento, el servicio de IA no esta configurado correctamente."

        try:
            # Check if we should show menu
            if self.should_show_menu(user_id):
                current_flow = config_service.get_current_flow()
                menu_config = config_service.get_menu_for_flow(current_flow)

                if menu_config:
                    self.user_menu_state[user_id] = True
                    return self.build_menu_message(menu_config)

            # Check if this is a menu selection
            current_flow = config_service.get_current_flow()
            flow_data = config_service.get_flow_data(current_flow)

            if flow_data and flow_data.get("has_menu") and flow_data.get("menu_config"):
                menu_response = self.handle_menu_selection(
                    user_id, user_message, flow_data["menu_config"]
                )
                if menu_response:
                    print("  Menu response selected")
                    return menu_response

            # Normal AI response
            if user_id not in self.conversations:
                self.conversations[user_id] = []
                print("  New conversation created")

            self.conversations[user_id].append({
                "role": "user",
                "content": user_message
            })

            # Keep last 20 messages
            if len(self.conversations[user_id]) > 20:
                self.conversations[user_id] = self.conversations[user_id][-20:]
                print("  History trimmed to 20 messages")

            print("  Sending to Grok API...")
            print("  Model: grok-4-fast-reasoning")
            print(f"  Messages in context: {len(self.conversations[user_id])}")

            current_prompt = self.system_prompt or config_service.get_system_prompt()

            completion = self.client.chat.completions.create(
                model="grok-4-fast-reasoning",
                messages=[
                    {"role": "system", "content": current_prompt},
                    *self.conversations[user_id]
                ],
                temperature=0.7,
                max_tokens=1000
            )

            print("  Response received from Grok")

            assistant_message = completion.choices[0].message.content
            print(f"  Response: {assistant_message[:100]}...")

            self.conversations[user_id].append({
                "role": "assistant",
                "content": assistant_message
            })

            return assistant_message

        except Exception as error:
            print(f"\nERROR IN GROK API:")
            print(f"  Message: {str(error)}")
            return "Disculpa, hubo un error tecnico. Puedes intentar de nuevo?"

    def clear_conversation(self, user_id: str) -> None:
        """Clear conversation history for a user."""
        self.conversations[user_id] = []
        self.user_menu_state[user_id] = False
        print(f"Conversation reset for: {user_id}")

    def get_conversation_history(self, user_id: str) -> List[Dict[str, str]]:
        """Get conversation history for a user."""
        return self.conversations.get(user_id, [])


# Singleton instance
grok_service = GrokService()
